File: Source/TWX26/Utility.pas
{
Copyright (C) 2005  Remco Mulder

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

For source notes please refer to Notes.txt
For license terms please refer to GPL.txt.

These files should be stored in the root of the compression you 
received this source in.
}
// This unit controls all utility functions used by the other units

unit
  Utility;

interface

uses
  Classes;

function GetSpace(I : Integer) : String;
function AsterixToEnter(S : String) : String;
function GetParameter(S : String; Parameter : Word) : String;
function GetParameterPos(S : String; Parameter : Integer) : Integer;
function Segment(I : Integer) : String;
function IsIn(S, X : String) : Boolean;
function StrToIntSafe(S : String) : Integer;
function StripChar(var S : String; C : Char) : String;
function ShortFilename(S : String) : String;
function StripFileExtension(S : string) : string;
function GetDirectory(S : String) : String;
function FetchScript(S : String; Include : Boolean) : String;
procedure Replace(var S : String; A, B : Char);
procedure CompleteFileName(var S : String; X : String);
function GetTelnetLogin(InStr : String) : String;
function StripChars(S : String) : String;
procedure FreeList(List : TList; MemSize : Integer);
procedure SetFileExtension(var Filename : String; Extension : String);
function WordWrap(S : string) : string;
function IsIpAddress(Addr : string) : Boolean;
function rdtsc : Int64;
procedure ReverseTList(var List : TList);
procedure Split(const Line : string; var Strings : TStringList; const Delimiters : string = ''); // default string parameter should be blank

implementation

uses
  SysUtils;

procedure SetFileExtension(var Filename : String; Extension : String);
begin
  // ensure the filename has an extention.  If not, make one
  if (Pos('.', Filename) = 0) then
    Filename := Filename + Extension;
end;

function GetSpace(I : Integer) : String;
var
  X : Integer;
begin
  Result := '';

  for X := 1 to I do
    Result := Result + ' ';
end;

function StripChars(S : String) : String;
var
  I : Integer;
begin
  Result := '';

  // remove all unusual characters from line
  for I := 1 to Length(S) do
    if (Byte(S[I]) >= 32) and (Byte(S[I]) < 127) then
      Result := Result + S[I];
end;

function GetTelnetLogin(InStr : String) : String;
var
  I,
  X    : Integer;
  Retn : String;
begin
  // get telnet commands from this line
  X := 0;
  Retn := '';

  for I := 1 to Length(InStr) do
  begin
    if (X > 0) then
      Dec(X);

    if (InStr[I] = #255) then
      X := 3;

    if (X > 0) then
      Retn := Retn + InStr[I];
  end;

  GetTelnetLogin := Retn;
end;

function Generate(Name, Base : String) : String;
var
  I    : Integer;
  Key,
  A,
  B    : Byte;
  Retn : String;
begin
  Retn := '';
  Key := $5A;
  B := $9F;

  for I := 1 to Length(Name) do
    Key := ((Key xor Byte(Name[I])) * I) xor 127 xor Byte(Name[2]);

  for I := 1 to Length(Base) do
  begin
    A := Byte(Base[I]);
    A := A xor Key xor B * I;
    B := A;

    A := A DIV 8;

    if (A < 10) then
      A := A + 48
    else
      A := A + 55;

    Retn := Retn + Char(A);
  end;

  Result := Retn;
end;

function FetchScript(S : String; Include : Boolean) : String;
var
  NameList : array[1..6] of string;
  Limit,
  I        : Integer;
begin
  I := 1;

  while (I <= 5) do
  begin
    NameList[I] := S;
    Inc(I, 2);
  end;

  I := 2;

  while (I <= 6) do
  begin
    NameList[I] := 'scripts\' + S;
    Inc(I, 2);
  end;

  CompleteFileName(NameList[1], 'ts');
  CompleteFileName(NameList[2], 'ts');
  CompleteFileName(NameList[3], 'cts');
  CompleteFileName(NameList[4], 'cts');
  CompleteFileName(NameList[5], 'inc');
  CompleteFileName(NameList[6], 'inc');

  if (Include) then
    Limit := 6
  else
    Limit := 4;

  Result := S;

  for I := 1 to Limit do
    if (FileExists(NameList[I])) then
    begin
      Result := NameList[I];
      Break;
    end;
end;

procedure CompleteFileName(var S : String; X : String);
var
  I : Integer;
begin
  // add an extention to the filename if there isn't one
  for I := length(S) downto 1 do
  begin
    if (S[I] = '.') then
      exit;
    if (S[I] = '\') then
      break;
  end;

  S := S + '.' + X;
end;

procedure Replace(var S : String; A, B : Char);
var
  I : Integer;
  N : String;
begin
  N := '';

  for I := 1 to Length(S) do
    if (S[I] = A) then
      N := N + B
    else
      N := N + S[I];

  S := N;
end;

function ShortFilename(S : String) : String;
var
  I : Integer;
begin
  // take the directories out of the filename
  for I := length(S) downto 1 do
    if (S[I] = '\') then
      break;

  ShortFilename := Copy(S, I + 1, length(S));
end;

function StripFileExtension(S : string) : string;
var
  I : Integer;
begin
  // take the extension out of the filename (if its there)
  for I := length(S) downto 1 do
    if (S[I] = '.') then
    begin
      Result := Copy(S, 1, I - 1);
      Exit;
    end;

  Result := S;
end;

function StrToIntSafe(S : String) : Integer;
var
  ErrorC,
  Value  : Integer;
begin
  Val(S, Value, ErrorC);

  if (ErrorC <> 0) then
    Value := 0;

  StrToIntSafe := Value;
end;

function AsterixToEnter(S : String) : String;
var
  X : String;
  I : Integer;
begin
  for I := 1 to length(S) do
    if (S[I] = '*') then
    begin
      X := X + chr(13);
      X := X + chr(10);
    end
    else
      X := X + S[I];

  AsterixToEnter := X;
end;

function Segment(I : Integer) : String;
var
  X, C : Integer;
  M, O : String;
begin
  // Moving from the right side, add commas every 3rd digit
  M := IntToStr(I);
  O := '';
  C := -1;

  for X := 0 to length(M) - 1 do
  begin
    Inc(C);
    if (C = 3) then
    begin
      C := 0;
      O := ',' + O;
    end;
    O := M[length(M) - X] + O;
  end;

  Segment := O;
end;

function GetDirectory(S : String) : String;
var
  I    : Integer;
  Retn : String;
begin
  // gets the directory out of the passed filename

  Retn := '';
  for I := Length(S) downto 1 do
  begin
    if (S[I] = '\') then
    begin
      Retn := Copy(S, 1, I - 1);
      Break;
    end;
  end;

  GetDirectory := Retn;
end;

function StripChar(var S : String; C : Char) : String;
var
  I : Integer;
  X : String;
begin
  // Remove character from string
  X := '';

  for I := 1 to length(S) do
    if (S[I] <> C) then X := X + S[I];

  S := X;
  StripChar := X;
end;

function IsIn(S, X : String) : Boolean;
begin
  if (S = '') then
    IsIn := TRUE
  else
  begin
    if (Pos(S, X) > 0) then
      IsIn := TRUE
    else
      IsIn := FALSE;
  end;
end;

function GetParameter(S : String; Parameter : Word) : String;
var
  I, P : Integer;
  retn : String;
  Last : Char;
begin
  // Get text parameter
  if (S = '') then
  begin
    GetParameter := '';
    Exit;
  end;

  P := 1;
  Last := ' ';
  retn := '';

  for I := 1 to length(S) do
  begin
    if (P = Parameter) and (S[I] <> ' ') and (S[I] <> #9) then
      retn := retn + S[I];

    if ((S[I] = ' ') or (S[I] = #9)) and (Last <> ' ') and (Last <> #9) then
    begin
      Inc(P);
      if (P > Parameter) then
        break;
    end;

    Last := S[I];
  end;

  GetParameter := retn;
end;

function GetParameterPos(S : String; Parameter : Integer) : Integer;
var
  I, P : Integer;
  Last : Char;
begin
  // Get text parameter
  P := 1;
  Last := ' ';

  for I := 1 to length(S) do
  begin
    if (S[I] = ' ') and (Last <> ' ') then
      Inc(P);

    if (P = Parameter) then
      break;

    Last := S[I];
  end;

  GetParameterPos := I + 1;
end;

procedure FreeList(List : TList; MemSize : Integer);
begin
  // remove all items and free list
  while (List.Count > 0) do
  begin
    FreeMem(List[0], MemSize);
    List.Delete(0);
  end;

  List.Free;
end;

function WordWrap(S : string) : string;
var
  I,
  J,
  Col : Integer;
begin
  // word-wrap value to max 60 characters
  Col := 1;
  for I := 1 to Length(S) do
  begin
    if (S[I] = #13) then
      Col := 0
    else
      Inc(Col);

    if (Col > 60) and (S[I] = ' ') then
    begin
      for J := I to Length(S) do
        if (S[J] <> ' ') then
        begin
          Insert(#13 + #10, S, J);
          Col := 0;
          Break;
        end;
    end;
  end;

  Result := S;
end;

function IsIpAddress(Addr : string) : Boolean;
var
  I,
  Dots : Integer;
begin
  Result := TRUE;
  Dots := 0;

  for I := 1 to Length(Addr) do
  begin
    if (Addr[I] = '.') then
      Inc(Dots)
    else if (Byte(Addr[I]) < 48) or (Byte(Addr[I]) > 57) then
    begin
      Result := FALSE;
      Break;
    end;
  end;

  if (Dots <> 3) then
    Result := FALSE;
end;

function rdtsc : Int64;
asm
  rdtsc
end;

procedure ReverseTList(var List : TList);
// This procedure flips a lists' order, primarily for course lists
var
  I,
  J : Integer;
begin
  J := List.Count div 2;
  for I := 1 to J do
    List.Exchange(I - 1, List.Count - I);
end;

procedure Split(const Line : string; var Strings : TStringList; const Delimiters : string = ''); // default string parameter should be blank
var
  Separators,
  WhiteSpace : set of Char; // set of Char
  I : Integer;
begin
  if (Delimiters <> '') then // Delimiters were specified
  begin
    Separators := [];
    for I := 1 to Length(Delimiters) do
      Separators := Separators + [Delimiters[I]];
  end
  else
    Separators := [#9, #32]; // Tab and Space, if omitted

  WhiteSpace := [];
  ExtractStrings(Separators, WhiteSpace, @Line[1], TStrings(Strings));
end;

end.

File: Source/TWX30/Source/Core/Utility.cs
/*
Copyright (C) 2005  Remco Mulder

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

For source notes please refer to Notes.txt
For license terms please refer to GPL.txt.

These files should be stored in the root of the compression you 
received this source in.
*/

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;

namespace TWXProxy.Core
{
    /// <summary>
    /// This class controls all utility functions used by the other units
    /// </summary>
    public static class Utility
    {
        public static string GetSpace(int count)
        {
            return new string(' ', count);
        }

        public static string AsterixToEnter(string s)
        {
            return s.Replace("*", "\r\n");
        }

        public static string GetParameter(string s, int parameter)
        {
            // Get text parameter
            if (string.IsNullOrEmpty(s))
                return string.Empty;

            int p = 1;
            char last = ' ';
            string result = string.Empty;

            for (int i = 0; i < s.Length; i++)
            {
                if (p == parameter && s[i] != ' ' && s[i] != '\t')
                    result += s[i];

                if ((s[i] == ' ' || s[i] == '\t') && last != ' ' && last != '\t')
                {
                    p++;
                    if (p > parameter)
                        break;
                }

                last = s[i];
            }

            return result;
        }

        public static int GetParameterPos(string s, int parameter)
        {
            // Get text parameter position
            int p = 1;
            char last = ' ';

            for (int i = 0; i < s.Length; i++)
            {
                if (s[i] == ' ' && last != ' ')
                    p++;

                if (p == parameter)
                    return i + 1;

                last = s[i];
            }

            return s.Length;
        }

        public static string Segment(int value)
        {
            // Moving from the right side, add commas every 3rd digit
            return value.ToString("N0");
        }

        public static bool IsIn(string s, string x)
        {
            if (string.IsNullOrEmpty(s))
                return true;

            return x.Contains(s);
        }

        public static int StrToIntSafe(string s)
        {
            if (int.TryParse(s, out int value))
                return value;
            return 0;
        }

        public static string StripChar(string s, char c)
        {
            return s.Replace(c.ToString(), string.Empty);
        }

        public static string ShortFilename(string s)
        {
            // take the directories out of the filename
            return Path.GetFileName(s);
        }

        public static string StripFileExtension(string s)
        {
            // take the extension out of the filename (if its there)
            return Path.GetFileNameWithoutExtension(s);
        }

        public static string GetDirectory(string s)
        {
            // gets the directory out of the passed filename
            return Path.GetDirectoryName(s) ?? string.Empty;
        }

        public static string FetchScript(string s, bool include)
        {
            string[] nameList = new string[6];
            int limit = include ? 6 : 4;

            // Initialize name variations
            for (int i = 0; i < 3; i++)
                nameList[i * 2] = s;

            for (int i = 0; i < 3; i++)
                nameList[i * 2 + 1] = Path.Combine("scripts", s);

            CompleteFileName(ref nameList[0], "ts");
            CompleteFileName(ref nameList[1], "ts");
            CompleteFileName(ref nameList[2], "cts");
            CompleteFileName(ref nameList[3], "cts");
            CompleteFileName(ref nameList[4], "inc");
            CompleteFileName(ref nameList[5], "inc");

            for (int i = 0; i < limit; i++)
            {
                if (File.Exists(nameList[i]))
                    return nameList[i];
            }

            return s;
        }

        public static void CompleteFileName(ref string s, string extension)
        {
            // add an extension to the filename if there isn't one
            if (!Path.HasExtension(s))
                s = s + "." + extension;
        }

        public static void Replace(ref string s, char a, char b)
        {
            s = s.Replace(a, b);
        }

        public static string NormalizePathSeparators(string path)
        {
            if (string.IsNullOrWhiteSpace(path))
                return path;

            return Path.DirectorySeparatorChar == '\\'
                ? path.Replace('/', Path.DirectorySeparatorChar)
                : path.Replace('\\', Path.DirectorySeparatorChar);
        }

        public static string ResolvePlatformPath(string path, string? baseDir = null)
        {
            if (string.IsNullOrWhiteSpace(path))
                return path;

            string normalized = NormalizePathSeparators(path);
            if (!Path.IsPathRooted(normalized))
            {
                string root = string.IsNullOrWhiteSpace(baseDir)
                    ? Directory.GetCurrentDirectory()
                    : baseDir;
                normalized = Path.Combine(root, normalized);
            }

            return normalized;
        }

        public static string? ResolveExistingPath(string path, string? baseDir = null)
        {
            return ResolveExistingPath(path, baseDir, requireFile: false, requireDirectory: false);
        }

        public static string? ResolveExistingFilePath(string path, string? baseDir = null)
        {
            return ResolveExistingPath(path, baseDir, requireFile: true, requireDirectory: false);
        }

        public static string? ResolveExistingDirectoryPath(string path, string? baseDir = null)
        {
            return ResolveExistingPath(path, baseDir, requireFile: false, requireDirectory: true);
        }

        private static string? ResolveExistingPath(
            string path,
            string? baseDir,
            bool requireFile,
            bool requireDirectory)
        {
            if (string.IsNullOrWhiteSpace(path))
                return null;

            string fullPath = Path.GetFullPath(ResolvePlatformPath(path, baseDir));
            if (PathMatchesKind(fullPath, requireFile, requireDirectory))
                return fullPath;

            string? root = Path.GetPathRoot(fullPath);
            if (string.IsNullOrEmpty(root))
                return null;

            string current = root;
            string relative = fullPath.Substring(root.Length);
            string[] parts = relative.Split(
                new[] { Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar },
                StringSplitOptions.RemoveEmptyEntries);

            foreach (string part in parts)
            {
                if (!Directory.Exists(current))
                    return null;

                string? match;
                try
                {
                    match = Directory.EnumerateFileSystemEntries(current)
                        .FirstOrDefault(entry => string.Equals(Path.GetFileName(entry), part, StringComparison.OrdinalIgnoreCase));
                }
                catch
                {
                    return null;
                }

                if (string.IsNullOrEmpty(match))
                    return null;

                current = match;
            }

            return PathMatchesKind(current, requireFile, requireDirectory) ? current : null;
        }

        private static bool PathMatchesKind(string path, bool requireFile, bool requireDirectory)
        {
            if (requireFile)
                return File.Exists(path);

            if (requireDirectory)
                return Directory.Exists(path);

            return File.Exists(path) || Directory.Exists(path);
        }

        public static string GetTelnetLogin(string inStr)
        {
            // get telnet commands from this line
            int x = 0;
            string result = string.Empty;

            for (int i = 0; i < inStr.Length; i++)
            {
                if (x > 0)
                    x--;

                if (inStr[i] == (char)255)
                    x = 3;

                if (x > 0)
                    result += inStr[i];
            }

            return result;
        }

        public static string StripChars(string s)
        {
            // remove all unusual characters from line
            string result = string.Empty;

            for (int i = 0; i < s.Length; i++)
            {
                if (s[i] >= 32 && s[i] < 127)
                    result += s[i];
            }

            return result;
        }

        public static void SetFileExtension(ref string filename, string extension)
        {
            // ensure the filename has an extension.  If not, make one
            if (!filename.Contains("."))
                filename = filename + extension;
        }

        public static string WordWrap(string s, int maxWidth = 60)
        {
            // word-wrap value to max characters
            int col = 1;
            char[] chars = s.ToCharArray();

            for (int i = 0; i < chars.Length; i++)
            {
                if (chars[i] == '\r')
                    col = 0;
                else
                    col++;

                if (col > maxWidth && chars[i] == ' ')
                {
                    for (int j = i; j < chars.Length; j++)
                    {
                        if (chars[j] != ' ')
                        {
                            s = s.Insert(j, "\r\n");
                            col = 0;
                            break;
                        }
                    }
                }
            }

            return s;
        }

        public static bool IsIpAddress(string addr)
        {
            int dots = 0;

            foreach (char c in addr)
            {
                if (c == '.')
                    dots++;
                else if (c < '0' || c > '9')
                    return false;
            }

            return dots == 3;
        }

        public static void ReverseTList<T>(List<T> list)
        {
            // This procedure flips a lists' order
            list.Reverse();
        }

        public static List<string> Split(string line, string? delimiters = null)
        {
            char[] separators = string.IsNullOrEmpty(delimiters)
                ? new[] { '\t', ' ' }
                : delimiters.ToCharArray();

            return line.Split(separators, StringSplitOptions.RemoveEmptyEntries).ToList();
        }

        public static string Generate(string name, string baseStr)
        {
            byte key = 0x5A;
            byte b = 0x9F;

            for (int i = 0; i < name.Length; i++)
            {
                key = (byte)(((key ^ (byte)name[i]) * (i + 1)) ^ 127 ^ (byte)name[Math.Min(1, name.Length - 1)]);
            }

            string result = string.Empty;
            for (int i = 0; i < baseStr.Length; i++)
            {
                byte a = (byte)baseStr[i];
                a = (byte)(a ^ key ^ (b * (i + 1)));
                b = a;

                a = (byte)(a / 8);

                if (a < 10)
                    a += 48;
                else
                    a += 55;

                result += (char)a;
            }

            return result;
        }
    }
}
